backend/src/services/pattern_extraction_service.py

```
@@ -1,134 +1,698 @@
+                return []
+            
+            # Get user's transactions - unpack the tuple
+            all_transactions, _, _ = list_user_transactions(user_id)
+            
+            # Score transactions by similarity 
+            scored_transactions = []
+            for t in all_transactions:
+                # Skip comparison with same transaction (simplified approach)
+                if hasattr(t, 'transaction_id') and hasattr(transaction, 'transaction_id'):
+                    try:
+                        if str(getattr(t, 'transaction_id')) == str(getattr(transaction, 'transaction_id')):
```

Ruff (0.11.9)

Do not call `getattr` with a constant attribute value. It is not any safer than normal property access.

Replace `getattr` with attribute access

(B009)